### ISSUE TITLE
TypeError of Sabre when using the Busy/Free functionality of calendar

### ISSUE THREAD

**Issue by torric** [nextcloud/server#9326]
### Steps to reproduce
1. Create a new event in personal calendar in Thunderbird.
2. Click on Invite Attendees (Teilnehmer einladen)
3. Bug message appears immediately in Nextcloud log. The same happens when other people try to invite me.

### Expected behaviour
Thunderbird should show my existing events.

### Actual behaviour
Bug message appears immediately in Nextcloud log. My Thunderbird or that of an other user does not show my events. My user account with the problem is the second one that was created after the installation of Nextcloud 10 but it is not an administrator. Newer users do not have the problem described here.

### Server configuration

**Operating system**: Ubuntu Server 16.04.4 LTS

**Web server:** Apache 2.4.18

**Database:** MariaDB 10.0.34

**PHP version:** 7.0.28

**Nextcloud version:** 13.0.2

**Updated from an older Nextcloud/ownCloud or fresh install:** Nextcloud 10

**Where did you install Nextcloud from:** zip

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
  - apporder: 0.4.1
  - bruteforcesettings: 1.0.3
  - calendar: 1.6.1
  - contacts: 2.1.3
  - dav: 1.4.6
  - federatedfilesharing: 1.3.1
  - files: 1.8.0
  - files_sharing: 1.5.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.1.0
  - nextcloud_announcements: 1.2.0
  - notifications: 2.1.2
  - oauth2: 1.1.0
  - password_policy: 1.3.0
  - polls: 0.8.1
  - provisioning_api: 1.3.0
  - serverinfo: 1.3.0
  - sharebymail: 1.3.0
  - spreed: 3.1.0
  - tasks: 0.9.6
  - theming: 1.4.1
  - twofactor_backupcodes: 1.2.3
  - updatenotification: 1.3.0
  - user_ldap: 1.3.1
  - workflowengine: 1.3.0

```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            ""
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "...",
        "dbtype": "mysql",
        "version": "13.0.2.1",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true,
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.local": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "ldapIgnoreNamingRules": false,
        "ldapProviderFactory": "\\OCA\\User_LDAP\\LDAPProviderFactory",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "587",
        "mail_smtpsecure": "tls",
        "defaultapp": "calendar",
        "session_lifetime": 600,
        "skeletondirectory": "",
        "maintenance": false,
        "loglevel": 1,
        "singleuser": false,
        "updater.release.channel": "stable",
        "theme": "",
        "updater.secret": "***REMOVED SENSITIVE VALUE***"
    }

```
</details>

**Are you using external storage, if yes which one:** no

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** ActiveDirectory

#### LDAP configuration (delete this part if not used)
<details>
<summary>LDAP config</summary>

```
+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------
| hasMemberOfFilterSupport      | 1
| hasPagedResultSupport         |
| homeFolderNamingRule          | attr:samaccountname
| lastJpegPhotoLookup           | 0
| ldapAgentName                 | cn=ldapquery,cn=Users,dc=...,dc=...,dc=de
| ldapAgentPassword             | ***
| ldapAttributesForGroupSearch  |
| ldapAttributesForUserSearch   | name;displayName
| ldapBackupHost                | ...de
| ldapBackupPort                | 389
| ldapBase                      | ou=...,dc=...,dc=...,dc=de
| ldapBaseGroups                | ou=...,dc=...,dc=...,dc=de
| ldapBaseUsers                 | ou=...,dc=...,dc=...,dc=de
| ldapCacheTTL                  | 600
| ldapConfigurationActive       | 1
| ldapDefaultPPolicyDN          |
| ldapDynamicGroupMemberURL     |
| ldapEmailAttribute            | mail
| ldapExperiencedAdmin          | 1
| ldapExpertUUIDGroupAttr       |
| ldapExpertUUIDUserAttr        | cn
| ldapExpertUsernameAttr        | samaccountname
| ldapGidNumber                 | gidNumber
| ldapGroupDisplayName          | cn
| ldapGroupFilter               | (|(cn=...)(cn=...)(cn=...)(cn=...)(cn=...))
| ldapGroupFilterGroups         | ...;...;...;...;...
| ldapGroupFilterMode           | 0
| ldapGroupFilterObjectclass    |
| ldapGroupMemberAssocAttr      | member
| ldapHost                      | ...de
| ldapIgnoreNamingRules         |
| ldapLoginFilter               | (&(& (mail=*) (objectclass=person))(|(samaccountname=%uid)(|(mailPrimaryAddress=%uid)(mail=%uid))))
| ldapLoginFilterAttributes     |
| ldapLoginFilterEmail          | 1
| ldapLoginFilterMode           | 0
| ldapLoginFilterUsername       | 1
| ldapNestedGroups              | 0
| ldapOverrideMainServer        |
| ldapPagingSize                | 500
| ldapPort                      | 389
| ldapQuotaAttribute            |
| ldapQuotaDefault              | 10 MB
| ldapTLS                       | 0
| ldapUserDisplayName           | name
| ldapUserDisplayName2          |
| ldapUserFilter                | (&(&(mail=*@...)(objectclass=person))(|(|(memberof=CN=...,OU=...,OU=...,DC=...,DC=...,DC=de) ... (primaryGroupID=1119)))) |
| ldapUserFilterGroups          |
| ldapUserFilterMode            | 0
| ldapUserFilterObjectclass     | person
| ldapUuidGroupAttribute        | auto
| ldapUuidUserAttribute         | auto
| turnOffCertCheck              | 0
| turnOnPasswordChange          | 0
| useMemberOfToDetectMembership | 1
+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------
```
</details>

### Client configuration
**Browser:** Thunderbird 52.7. 0 with Lightning 5.4.7  (Firefox 59.0.2)

**Operating system:** Windows 7 Pro x64

#### Web server error log
<details>
<summary>Web server error log</summary>

```
Insert your webserver log here
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
{"reqId":"l0CxV3UF0gaRjnJuACk2","level":3,"time":"2018-04-26T14:15:50+00:00","remoteAddr":"IP_2","user":"User2","app":"remote","method":"POST","url":"\/remote.php\/dav\/calendars\/User2\/outbox\/","message":"Exception: {\"Exception\":\"TypeError\",\"Message\":\"Argument 2 passed to OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Principal::OCA\\\\DAV\\\\Connector\\\\Sabre\\\\{closure}() must implement interface OCP\\\\IUser, null given\",\"Code\":0,\"Trace\":\"#0 [internal function]: OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Principal->OCA\\\\DAV\\\\Connector\\\\Sabre\\\\{closure}(*** sensitive parameters replaced ***)\\n#1 \\\/var\\\/www\\\/nextcloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Principal.php(242): array_reduce(Array, Object(Closure), Array)\\n#2 \\\/var\\\/www\\\/nextcloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Principal.php(280): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Principal->searchUserPrincipals(Array, 'allof')\\n#3 \\\/var\\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAVACL\\\/AbstractPrincipalCollection.php(148): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Principal->searchPrincipals('principals\\\/user...', Array, 'allof')\\n#4 \\\/var\\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAVACL\\\/Plugin.php(777): Sabre\\\\DAVACL\\\\AbstractPrincipalCollection->searchPrincipals(Array, 'allof')\\n#5 \\\/var\\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/CalDAV\\\/Schedule\\\/Plugin.php(910): Sabre\\\\DAVACL\\\\Plugin->principalSearch(Array, Array)\\n#6 \\\/var\\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/CalDAV\\\/Schedule\\\/Plugin.php(836): Sabre\\\\CalDAV\\\\Schedule\\\\Plugin->getFreeBusyForEmail('prename.surname...', Object(DateTimeImmutable), Object(DateTimeImmutable), Object(Sabre\\\\VObject\\\\Component\\\\VCalendar))\\n#7 \\\/var\\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/CalDAV\\\/Schedule\\\/Plugin.php(774): Sabre\\\\CalDAV\\\\Schedule\\\\Plugin->handleFreeBusyRequest(Object(Sabre\\\\CalDAV\\\\Schedule\\\\Outbox), Object(Sabre\\\\VObject\\\\Component\\\\VCalendar), Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#8 \\\/var\\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/CalDAV\\\/Schedule\\\/Plugin.php(187): Sabre\\\\CalDAV\\\\Schedule\\\\Plugin->outboxRequest(Object(Sabre\\\\CalDAV\\\\Schedule\\\\Outbox), Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#9 [internal function]: Sabre\\\\CalDAV\\\\Schedule\\\\Plugin->httpPost(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#10 \\\/var\\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#11 \\\/var\\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:POST', Array)\\n#12 \\\/var\\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#13 \\\/var\\\/www\\\/nextcloud\\\/apps\\\/dav\\\/lib\\\/Server.php(283): Sabre\\\\DAV\\\\Server->exec()\\n#14 \\\/var\\\/www\\\/nextcloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(35): OCA\\\\DAV\\\\Server->exec()\\n#15 \\\/var\\\/www\\\/nextcloud\\\/remote.php(164): require_once('\\\/var\\\/www\\\/nextcl...')\\n#16 {main}\",\"File\":\"\\\/var\\\/www\\\/nextcloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Principal.php\",\"Line\":231}","userAgent":"Mozilla\/5.0 (Windows NT 10.0; WOW64; rv:52.0) Gecko\/20100101 Thunderbird\/52.7.0 Lightning\/5.4","version":"13.0.1.1"}


TypeError: Argument 2 passed to OCA\DAV\Connector\Sabre\Principal::OCA\DAV\Connector\Sabre\{closure}() must implement interface OCP\IUser, null given

    [internal function] OCA\DAV\Connector\Sabre\Principal->OCA\DAV\Connector\Sabre\{closure}(*** sensitive parameters replaced ***)
    /var/www/nextcloud/apps/dav/lib/Connector/Sabre/Principal.php - line 242: array_reduce(Array, Object(Closure), Array)
    /var/www/nextcloud/apps/dav/lib/Connector/Sabre/Principal.php - line 280: OCA\DAV\Connector\Sabre\Principal->searchUserPrincipals(Array, 'allof')
    /var/www/nextcloud/3rdparty/sabre/dav/lib/DAVACL/AbstractPrincipalCollection.php - line 148: OCA\DAV\Connector\Sabre\Principal->searchPrincipals('principals/user...', Array, 'allof')
    /var/www/nextcloud/3rdparty/sabre/dav/lib/DAVACL/Plugin.php - line 777: Sabre\DAVACL\AbstractPrincipalCollection->searchPrincipals(Array, 'allof')
    /var/www/nextcloud/3rdparty/sabre/dav/lib/CalDAV/Schedule/Plugin.php - line 910: Sabre\DAVACL\Plugin->principalSearch(Array, Array)
    /var/www/nextcloud/3rdparty/sabre/dav/lib/CalDAV/Schedule/Plugin.php - line 836: Sabre\CalDAV\Schedule\Plugin->getFreeBusyForEmail('prename.surname...', Object(DateTimeImmutable), Object(DateTimeImmutable), Object(Sabre\VObject\Component\VCalendar))
    /var/www/nextcloud/3rdparty/sabre/dav/lib/CalDAV/Schedule/Plugin.php - line 774: Sabre\CalDAV\Schedule\Plugin->handleFreeBusyRequest(Object(Sabre\CalDAV\Schedule\Outbox), Object(Sabre\VObject\Component\VCalendar), Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))
    /var/www/nextcloud/3rdparty/sabre/dav/lib/CalDAV/Schedule/Plugin.php - line 187: Sabre\CalDAV\Schedule\Plugin->outboxRequest(Object(Sabre\CalDAV\Schedule\Outbox), Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))
    [internal function] Sabre\CalDAV\Schedule\Plugin->httpPost(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))
    /var/www/nextcloud/3rdparty/sabre/event/lib/EventEmitterTrait.php - line 105: call_user_func_array(Array, Array)
    /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php - line 479: Sabre\Event\EventEmitter->emit('method POST', Array)
    /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php - line 254: Sabre\DAV\Server->invokeMethod(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))
    /var/www/nextcloud/apps/dav/lib/Server.php - line 283: Sabre\DAV\Server->exec()
    /var/www/nextcloud/apps/dav/appinfo/v2/remote.php - line 35: OCA\DAV\Server->exec()
    /var/www/nextcloud/remote.php - line 164: require_once('/var/www/nextcl...')
    {main}
```
</details>

**Comment by MorrisJobke** [comment]
cc @georgehrke

**Comment by ChristophWurst** [comment]
@georgehrke :ping_pong:

**Comment by georgehrke** [comment]
@MorrisJobke @ChristophWurst Obviously I could just remove the `IUser` type hint and check inside the closure whether `$user` implements `IUser`, but isn't the actual issue inside `IUserManager::getByEmail` here?

Is it the expected behaviour that it returns a mix of `IUser` and `null`?

https://github.com/nextcloud/server/blob/v13.0.2/apps/dav/lib/Connector/Sabre/Principal.php#L231L242

**Comment by ChristophWurst** [comment]
> Is it the expected behaviour that it returns a mix of `IUser` and `null`?

Does it? From looking at the code I'd say it always returns an array of `IUser`s https://github.com/nextcloud/server/blob/0bfe3da6643f6add02af0e1c355e2aeae9cece42/lib/private/User/Manager.php#L586-L597

**Comment by georgehrke** [comment]
I’d expect so too, but `$users` is coming straight from `IUserManager::getByEmail`

**Comment by ChristophWurst** [comment]
Could you attach a debugger to trace down where the string values are coming from?

**Comment by georgehrke** [comment]
@blizzz Is it possible, that the LDAP backend does not delete the email from `oc_preferences` if a user is deleted?

**Comment by blizzz** [comment]
> @blizzz Is it possible, that the LDAP backend does not delete the email from oc_preferences if a user is deleted?

It should be with the standard deletion process, not specific to LDAP. Provided, that triggering the deletion of the user in questions was kicked off, indeed, as it does not happen automatically: https://docs.nextcloud.com/server/14/admin_manual/configuration_user/user_auth_ldap_cleanup.html

**Comment by MorrisJobke** [comment]
Fix is in #11830
